### PR TITLE
feat: add connection resilience with XState health monitor

### LINKED ISSUE #364: Bug: MCP Server startup blocked when GitLab is unreachable
### Description / 问题描述

**English:**
When GitLab is unreachable (network issues, VPN disconnected, GitLab down), MCP Server startup is blocked during `connectionManager.initialize()`. This prevents `tools/list` from returning, causing the MCP client to timeout or fail to load tools.

**中文:**
当 GitLab 不可达（网络问题、VPN 断开、GitLab 宕机）时，MCP Server 在启动过程中被 `connectionManager.initialize()` 阻塞，导致 `tools/list` 无法返回，MCP 客户端超时或工具列表加载失败。

---

### Root Cause / 根因

**English:**
In `handlers.ts`, `connectionManager.initialize()` is called without a timeout during server setup. If GitLab is slow or unreachable, the initialization hangs indefinitely.

**中文:**
在 `handlers.ts` 中，`connectionManager.initialize()` 在服务启动时被调用，但没有设置超时。当 GitLab 响应慢或不可达时，初始化会无限期挂起。

---

### Proposed Fix / 修复方案

**English:**
Add an initialization timeout with `Promise.race`:

```typescript
// config.ts
const parsedInitTimeoutMs = parseInt(process.env.GITLAB_INIT_TIMEOUT_MS ?? "5000", 10);
export const INIT_TIMEOUT_MS =
  Number.isFinite(parsedInitTimeoutMs) && parsedInitTimeoutMs > 0 
    ? parsedInitTimeoutMs : 5000;

// handlers.ts
const initPromise = connectionManager.initialize();
const timeoutPromise = new Promise<never>((_, reject) => {
  setTimeout(
    () => reject(new Error(`Connection initialization timed out after ${INIT_TIMEOUT_MS}ms`)),
    INIT_TIMEOUT_MS
  );
});
await Promise.race([initPromise, timeoutPromise]);
```

**中文:**
添加初始化超时保护：
- 新增 `INIT_TIMEOUT_MS` 配置（默认 5 秒）
- 使用 `Promise.race` 包装初始化操作
- 超时后继续启动，工具列表立即可用，实际调用时再处理连接失败

---

### Benefits / 效果

**English:**
- Server startup completes within 5 seconds regardless of GitLab availability
- Tools list is immediately available
- Actual tool calls handle connection failures gracefully

**中文:**
- 无论 GitLab 是否可用，服务启动最多 5 秒完成
- 工具列表立即可用
- 实际工具调用时优雅处理连接失败

---

### Affected Files / 相关文件

- `src/config.ts` — add `INIT_TIMEOUT_MS` config
- `src/handlers.ts` — wrap `initialize()` with timeout

---

### Environment / 环境

- GitLab MCP version: v6.62.1
- Transport: stdio / HTTP

### LINKED ISSUE #374: fix: rehydrate SchemaIntrospector widget cache on introspection cache hit
## Summary

When `ConnectionManager` restores introspection data from the `introspectionCache` or `InstanceRegistry` cache, it copies `schemaInfo` onto the per-URL `InstanceState` but never feeds that data back into the `SchemaIntrospector` instance. As a result, `isWidgetAvailable()` — which delegates to `state.schemaIntrospector.isWidgetTypeAvailable(...)` — can return `false` for widget types that are already present in `state.schemaInfo.workItemWidgetTypes`.

This is pre-existing behavior (same pattern before PR #369) and was surfaced during the review of #369.

## Affected code

`src/services/ConnectionManager.ts` — cache-hit branches in:
- `doInitialize()` (introspectionCache hit)
- `doIntrospection()` (InstanceRegistry hit and legacy introspectionCache hit)
- `isWidgetAvailable()` helper (reads from schemaIntrospector instead of schemaInfo)

## Minimal fix

**Option A — derive widget availability directly from `schemaInfo`:**

```diff
  public isWidgetAvailable(widgetType: string): boolean {
    const url = this.currentInstanceUrl;
    if (!url) return false;
    const state = this.instances.get(url);
-   if (!state?.schemaIntrospector) return false;
-   return state.schemaIntrospector.isWidgetTypeAvailable(widgetType);
+   return state?.schemaInfo?.workItemWidgetTypes.includes(widgetType) ?? false;
  }
```

**Option B — rehydrate `SchemaIntrospector` after every cache restore** (each cache-hit branch).

A small regression test asserting `isWidgetAvailable()` returns `true` after a cache restore would prevent future regressions.

## References

- Raised in PR #369: https://github.com/structured-world/gitlab-mcp/pull/369
- Review comment: https://github.com/structured-world/gitlab-mcp/pull/369#discussion_r3014999994
- Requested by @polaz

### PR DESCRIPTION
**Pull request by polaz**
## Summary

- Add `HealthMonitor` service with XState v5 state machine for per-instance connection lifecycle
- Server startup completes within bounded time (default 5s) regardless of GitLab availability
- Automatic reconnection with exponential backoff and jitter when GitLab becomes unreachable
- Dynamic tool list reduction in disconnected mode (only `manage_context` for diagnostics)
- Per-URL `InstanceState` in `ConnectionManager` — each URL gets isolated client, introspection, and init state

## Technical Details

**State machine:** `connecting → healthy | degraded | disconnected | failed`
- `healthy`: Full introspection done (version + schema), all tools available
- `degraded`: GitLab reachable but introspection incomplete (REST-only or OAuth-deferred — known version but no schema), tools available
- `disconnected`: GitLab unreachable (transient), only context tools exposed, auto-reconnect in background
- `failed`: Auth/config error (permanent), no auto-reconnect, requires manual `forceReconnect`

**Per-URL ConnectionManager (`Map<string, InstanceState>`):**
- Each `initialize(url)` creates isolated state — no shared mutable fields to overwrite
- No-arg `getClient()` / `getInstanceInfo()` default to `currentInstanceUrl` for backward compat
- Explicit `getClient(url)` / `isConnected(url)` for per-URL access
- `getInstanceClient(url)` throws for unknown URLs instead of silently routing to wrong instance
- Failed inits delete their state entry (identity-guarded to avoid clobbering concurrent retries)
- `currentInstanceUrl` updated only after successful init
- `clearInflight(url)` for HealthMonitor to release hung init promises after timeout
- `latestRequestedUrl` guard prevents stale init from overwriting `currentInstanceUrl`
- `refreshCache(instanceUrl?)` resolves tier/version/scopes for the correct instance
- Bootstrap catch returns structured `CONNECTION_FAILED` instead of generic `"Bad Request"`
- Blocked disconnected calls recorded in condensed-mode error stats
- `normalizeInstanceUrl()` strips trailing slashes and API path suffixes for consistent Map keys
- `clearInflight(url)` clears both `initializePromises` and `introspectionPromises`
- Health filtering separated from `toolLookupCache` — lookup always has all tools,
  `getAllToolDefinitions()` applies unreachable-mode reduction for `tools/list`
- `isToolAvailableForInstance()` uses pre-fetched instanceInfo (not zero-arg singleton)
- Degraded state verified via `quickHealthCheck()` before committing (prevents false reachable)
- `connecting` state treated as optimistic for tools/list (full list, not context-only)
- REST-only tokens (no GraphQL access) skip `ensureIntrospected()` entirely
- Timeout/unreachable tool lists not cached (rebuilt each call to prevent stale state)
- `reinitialize(url)` clears both old and new instance state

**Degraded detection:**
- Checks both `version !== 'unknown'` AND `getSchemaInfo(url)` presence via `hasSchemaInfo()` helper
- REST-only / OAuth-deferred modes (real version, no schema) correctly report `degraded`

**Reconnect backoff:** `min(BASE_DELAY × 2^attempt, MAX_DELAY)` with ±10% jitter

**Error classification:**
- **transient** (ECONNREFUSED, timeouts, 5xx) → `disconnected`, auto-reconnect
- **auth** (401, 403) at startup → `failed`, no auto-reconnect
- **permanent** (400, 404, config) → `failed`, no auto-reconnect
- Runtime auth errors from tool calls do not affect health state (#370)

**Handler integration:**
- `healthMonitorStartup` shared promise — concurrent sessions await same init (resets on failure)
- `requestInstanceUrl` captured before `Promise.race` — timeout reports against correct URL
- `effectiveInstanceUrl` resolved via OAuth context → `getCurrentInstanceUrl()` → `GITLAB_BASE_URL`
- Disconnected `manage_context` bypass includes condensed-mode request tracking

## Known Limitations

- Runtime token revocation (401/403 during tool calls) doesn't transition to `failed` — #370
- No-arg callers in `ToolAvailability` / `WidgetAvailability` resolve against `currentInstanceUrl`, not per-request OAuth context — noted with TODO for follow-up
- HealthMonitor tracks only the configured instance URL; OAuth multi-instance URLs assumed reachable

## New Environment Variables

| Variable | Default | Description |
|----------|---------|-------------|
| `GITLAB_INIT_TIMEOUT_MS` | `5000` | Max startup wait for GitLab |
| `GITLAB_RECONNECT_BASE_DELAY_MS` | `5000` | Initial reconnect delay |
| `GITLAB_RECONNECT_MAX_DELAY_MS` | `60000` | Maximum reconnect delay |
| `GITLAB_HEALTH_CHECK_INTERVAL_MS` | `60000` | Health check interval when connected |
| `GITLAB_FAILURE_THRESHOLD` | `3` | Consecutive failures before disconnecting |

## Test Plan

**Automated (4817 unit + 398 integration = 5215 total):**
- [x] 4817 unit tests pass (147 suites) — covers all state machine transitions,
  error classification, per-URL isolation, cache identity guards, timeout behavior
- [x] 398 integration tests pass (29 suites) — real GitLab API calls with data lifecycle
- [x] Build passes (`yarn build`)
- [x] Lint passes (`yarn lint`, 0 errors)

**Manual E2E (SSE server, `.env.test` credentials):**
- [x] Start with valid GitLab token → `connecting → healthy` (18.8.3-ee ultimate,
  30 widget types, 185 schema types), all tools available, `tools/list_changed` broadcast
- [x] Start with unreachable hostname → `connecting → degraded` (fallback schema,
  version=unknown), server starts within 5s, tools available with reduced introspection
- [x] Start with invalid token → `connecting → degraded` (401 on introspection,
  fallback schema), auth errors surface during tool calls → transitions to `failed`
  after consecutive failures (covered by 81 HealthMonitor unit tests)
- [x] Disconnected → auto-reconnect with exponential backoff (unit test: line 522)
- [x] Failed state → no auto-reconnect, manual `forceReconnect` required (unit test: line 240)
- [x] Tool timeout → `timedOut` flag prevents late health updates (unit test + handler test)
- [x] Non-idempotent ops skip `Promise.race` (handler test: skip-race for mutations)

Closes #364, Closes #374

### CODE REVIEW

File: src/registry-manager.ts [hunk 1]
```diff
@@ -289,6 +290,22 @@ class RegistryManager {
...
+    // only expose context tools (manage_context for diagnostics like whoami)
+    let disconnectedMode = false;
+    try {
+      const healthMonitor = HealthMonitor.getInstance();
+      if (
+        healthMonitor.getMonitoredInstances().length > 0 &&
+        !healthMonitor.isAnyInstanceHealthy()
+      ) {
+        disconnectedMode = true;
+        logDebug('Building tool cache in disconnected mode — only context tools available');
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/registry-manager.ts]: @coderabbitai Fixed — connecting state now treated as reachable in isAnyInstanceHealthy(), preventing tools from being stripped during startup.

File: src/services/HealthMonitor.ts [hunk 2]
```diff
@@ -0,0 +1,708 @@
...
+    HealthMonitor.instance ??= new HealthMonitor();
+    return HealthMonitor.instance;
+  }
+
+  /**
+   * Register a callback for connection state changes.
+   */
+  public onStateChange(callback: StateChangeCallback): void {
+    this.stateChangeCallbacks.push(callback);
+  }
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/services/HealthMonitor.ts]: @coderabbitai Added code comment explaining single-registration pattern. Callbacks cleared on shutdown().

File: src/services/HealthMonitor.ts [hunk 3]
```diff
@@ -0,0 +1,753 @@
...
+
+    // If already initialized, verify with a health check and derive degraded from instance info
+    if (connectionManager.isConnected()) {
+      const healthy = await quickHealthCheck(input.instanceUrl);
+      if (!healthy) {
+        throw new Error(`Health check failed for ${input.instanceUrl}`);
+      }
+      try {
+        const info = connectionManager.getInstanceInfo();
+        return { degraded: info.version === 'unknown' };
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/services/HealthMonitor.ts]: @coderabbitai Acknowledged — added code comment explaining this is a known singleton limitation. Single-instance mode uses this path; multi-instance OAuth uses InstanceRegistry for per-instance state.

File: tests/unit/services/HealthMonitor.test.ts [hunk 4]
```diff
@@ -0,0 +1,898 @@
...
+  it('should classify health check failed as transient', () => {
+    const error = new Error('Health check failed for https://gitlab.example.com');
+    expect(classifyError(error)).toBe('transient');
+  });
+
+  it('should classify initialization timeout as transient', () => {
+    const error = new Error('Initialization timeout after 5000ms');
+    expect(classifyError(error)).toBe('transient');
+  });
+});
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/unit/services/HealthMonitor.test.ts]: @coderabbitai Won't merge — the two classifyError test blocks test different groups (basic vs edge cases) and are readable as separate describes.

File: src/services/HealthMonitor.ts [hunk 5]
```diff
@@ -0,0 +1,764 @@
...
+   */
+  private extractState(snapshot: SnapshotFrom<typeof connectionMachine>): ConnectionState {
+    const value = snapshot.value;
+    if (typeof value === 'string') {
+      return value as ConnectionState;
+    }
+    // Compound state: { healthy: 'idle' } or { healthy: 'checking' }
+    const topLevel = Object.keys(value)[0];
+    return topLevel as ConnectionState;
+  }
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/services/HealthMonitor.ts]: @coderabbitai Won't change — this machine has no parallel states, and adding defensive handling for a hypothetical future change adds complexity without value. The type assertion is safe for the current sequential state machine.

File: src/services/HealthMonitor.ts [hunk 6]
```diff
@@ -0,0 +1,813 @@
...
+    try {
+      const info = connectionManager.getInstanceInfo(input.instanceUrl);
+      const hasSchema = (() => {
+        try {
+          connectionManager.getSchemaInfo(input.instanceUrl);
+          return true;
+        } catch {
+          return false;
+        }
+      })();
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/services/HealthMonitor.ts]: @coderabbitai Fixed — replaced last IIFE with `hasSchemaInfo()` helper call.

**Comment by polaz** [comment]
@coderabbitai Re: review body items from latest review round:

**Duplicate: `isConnected()` still isn't a safe health signal (ConnectionManager.ts:439-446)**
→ **STALE.** `isConnected(url)` is now per-URL — checks `state.isInitialized` from `Map<string, InstanceState>`, not a singleton flag. JSDoc at lines 504-510 documents this is an initialization flag, not a connectivity check. HealthMonitor uses `hasSchemaInfo()` for degraded detection.

**Outside-diff: Connection bootstrap failures never reach the health monitor (handlers.ts:457-472)**
→ **ALREADY FIXED.** Lines 438-442 now catch bootstrap failures (`getClient`, `ensureIntrospected`, `initialize`) and call `healthMonitor.reportError(effectiveInstanceUrl, initError)`. The outer `handlerWork` try/catch wraps the entire post-gate section.

**Also from latest duplicate review (screenshot):**
- **`belowThreshold` unused guard** → FIXED in 0c4bd64 (removed dead code)
- **`currentInstanceUrl` rebind after detached init** → FIXED in 4ebc69f (line 87: `|| !this.currentInstanceUrl` fallback)

File: src/services/HealthMonitor.ts [hunk 7]
```diff
@@ -0,0 +1,820 @@
...
+      isDegraded = true;
+    }
+
+    if (isDegraded) {
+      // Verify reachability — OAuth/REST-only init can succeed with fallback
+      // data even when GitLab is unreachable. Throwing here lands in disconnected.
+      const reachable = await quickHealthCheck(input.instanceUrl);
+      if (!reachable) {
+        throw new Error('Instance unreachable despite successful init (degraded probe failed)');
+      }
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/services/HealthMonitor.ts]: @coderabbitai Fixed — degraded probe now throws `Health check failed: ...` matching transient pattern in classifyError().

File: src/handlers.ts [hunk 8]
```diff
@@ -120,22 +147,73 @@ function toStructuredError(
...
+                  RegistryManager.getInstance().refreshCache(instanceUrl);
+
+                  const { getSessionManager } = await import('./session-manager');
+                  await getSessionManager().broadcastToolsListChanged();
+
+                  logInfo('Tool list updated after connection state change', { from, to });
+                } catch (error) {
+                  logWarn('Failed to broadcast tools/list_changed after state change', {
+                    err: error as Error,
+                  });
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/handlers.ts]: @coderabbitai Already fixed in this push — `instanceUrl` added to both log lines (191 and 492).

File: src/config.ts [hunk 9]
```diff
@@ -391,6 +391,44 @@ export const HANDLER_TIMEOUT_MS =
...
+  process.env.GITLAB_HEALTH_CHECK_INTERVAL_MS ?? '60000',
+  10,
+);
+export const HEALTH_CHECK_INTERVAL_MS =
+  Number.isFinite(parsedHealthCheckInterval) && parsedHealthCheckInterval > 0
+    ? parsedHealthCheckInterval
+    : 60000;
+
+// Consecutive transient failures before transitioning to DISCONNECTED
+const parsedFailureThreshold = parseInt(process.env.GITLAB_FAILURE_THRESHOLD ?? '3', 10);
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/config.ts]: @coderabbitai Fixed — all `parseInt` calls replaced with `Number.parseInt`.

File: src/services/ConnectionManager.ts [hunk 10]
```diff
@@ -483,12 +614,14 @@ export class ConnectionManager {
...
+    const url = this.currentInstanceUrl;
+    const state = url ? this.instances.get(url) : undefined;
+    const previousScopes = state?.tokenScopeInfo?.scopes ?? [];
+    const previousHasGraphQL = state?.tokenScopeInfo?.hasGraphQLAccess ?? false;
+    const previousHasWrite = state?.tokenScopeInfo?.hasWriteAccess ?? false;
 
-    // Re-detect token scopes
-    const newScopeInfo = await detectTokenScopes();
+    // Re-detect token scopes against the current instance URL
+    const newScopeInfo = await detectTokenScopes(url ?? undefined);
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/services/ConnectionManager.ts]: @coderabbitai Fixed — `refreshTokenScopes()` now early-returns `false` when `currentInstanceUrl` or state is null.

File: tests/unit/handlers.test.ts [hunk 11]
```diff
@@ -422,10 +471,198 @@ describe('handlers', () => {
...
+
+      await callToolHandler({
+        params: {
+          name: 'test_tool',
+          arguments: {},
+        },
+      });
+
+      expect(mockHealthMonitor.reportError).toHaveBeenCalled();
+    });
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/unit/handlers.test.ts]: @coderabbitai Fixed — `reportSuccess` and `reportError` assertions now verify the per-instance URL argument.

File: tests/unit/services/HealthMonitor.test.ts [hunk 12]
```diff
@@ -0,0 +1,1042 @@
...
+    /** Set up mocks for "already connected" scenario */
+    function setupConnectedState(
+      version: string,
+      tier: string,
+      fetchStatus: { status: number; ok: boolean } = { status: 200, ok: true },
+    ): void {
+      mockIsConnected.mockReturnValue(true);
+      mockFetch.mockResolvedValue(fetchStatus);
+      mockGetInstanceInfo.mockReturnValue({ version, tier });
+    }
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/unit/services/HealthMonitor.test.ts]: @coderabbitai Fixed — hoisted setupConnectedState to outer describe scope, removed both inner copies.
